File: leetcode/leetcode_p152/code_152.py
# 力扣第152题，乘积最大子数组
"""
给你一个整数数组nums，请你找出数组中乘积最大的连续子数组(该子数组中至少包含一个数字)，
并返回该子数组所对应的乘积。
"""


class Solution:
    def maxProduct(self, nums):
        min_list = [nums[0]]
        max_list = [nums[0]]

        for num in nums[1:]:
            max_list.append(max([num, min_list[-1]*num, max_list[-1]*num]))
            min_list.append(min([num, min_list[-1]*num, max_list[-2]*num]))
        
        return max(max_list)


class Solution_01:
    def maxProduct(self, nums):
        result = nums[0]
        min_value = nums[0]
        max_value = nums[0]

        for num in nums[1:]:
            temp = min_value
            min_value = min([num, min_value * num, max_value * num])
            max_value = max([num, temp*num, max_value * num])
            if max_value > result:
                result = max_value
        
        return result


if __name__ == "__main__":
    # nums = [2, 3, -2, 4, -1]
    nums = [-4,-3,-2]
    sl1 = Solution_01()
    result1 = sl1.maxProduct(nums)
    print('result:', result1)

File: leetcode/leetcode_p152/code_152.cpp
// 力扣第152题，乘积最大子数组
/*
给你一个整数数组nums，请你找出数组中乘积最大的连续子数组(该子数组中至少包含一个数字)，
并返回该子数组所对应的乘积。
*/
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;


class Solution{
public:
    int maxProduct(vector<int>& nums){
        int result = nums[0];
        int max_value = result;
        int min_value = result;

        for (auto num=nums.begin()+1; num != nums.end(); num++){
            int temp = min_value;
            min_value = min(*num, min((*num)*temp, (*num)*max_value));
            max_value = max(*num, max((*num)*max_value, (*num)*temp));
            if (max_value > result)
                result = max_value;
        }
        return result;
    }
};


int main()
{
    vector<int> num = {3, -2, 4, -1};
    Solution sl;
    int result = sl.maxProduct(num);
    cout << "result = " << result << endl;

    cin.get();
    return 0;
}